File: app/src/main/java/io/inabsentia/superhangman/adapter/PickWordAdapter.java
package io.inabsentia.superhangman.adapter;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.ArrayAdapter;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.List;

import io.inabsentia.superhangman.R;

public class PickWordAdapter extends ArrayAdapter<String> {

    private List<String> words;
    private Context mContext;

    /*
     * ViewHolder that encapsulates the objects
     * that are used in the list view.
     */
    private static class ViewHolder {
        private TextView tvPos, tvWord;
        private ImageView image;
    }

    public PickWordAdapter(List<String> words, @NonNull Context context) {
        super(context, R.layout.pick_word_item, words);
        this.words = words;
        this.mContext = context;
    }

    private int lastPosition = -1;

    @NonNull
    @Override
    public View getView(int position, @Nullable View convertView, @NonNull ViewGroup parent) {

        final View view;
        ViewHolder viewHolder;

        final String word = words.get(position);

        if (convertView == null) {
            viewHolder = new ViewHolder();
            LayoutInflater inflater = LayoutInflater.from(getContext());
            convertView = inflater.inflate(R.layout.pick_word_item, parent, false);

            viewHolder.tvPos = convertView.findViewById(R.id.position_view);
            viewHolder.tvWord = convertView.findViewById(R.id.word_view);
            viewHolder.image = convertView.findViewById(R.id.image_view);

            view = convertView;
            convertView.setTag(viewHolder);
        } else {
            viewHolder = (ViewHolder) convertView.getTag();
            view = convertView;
        }

        Animation animation = AnimationUtils.loadAnimation(mContext, (position > lastPosition) ? R.anim.up_from_bottom : R.anim.down_from_top);
        view.startAnimation(animation);
        lastPosition = position;

        viewHolder.tvPos.setText(String.valueOf(position));
        viewHolder.tvWord.setText(word);

        return convertView;
    }

}

File: app/src/main/java/io/inabsentia/superhangman/adapter/PickWordAdapter.kt
package io.inabsentia.superhangman.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.AnimationUtils
import android.widget.ArrayAdapter
import android.widget.ImageView
import android.widget.TextView
import io.inabsentia.superhangman.R

class PickWordAdapter(private val words: List<String>, private val mContext: Context) : ArrayAdapter<String>(mContext, R.layout.pick_word_item, words) {

    private var lastPosition = -1

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        var mConvertView = convertView
        val view: View
        val viewHolder: ViewHolder

        val word = words[position]

        if (mConvertView == null) {
            viewHolder = ViewHolder()
            val inflater = LayoutInflater.from(context)
            mConvertView = inflater.inflate(R.layout.pick_word_item, parent, false)

            viewHolder.tvPos = mConvertView!!.findViewById(R.id.position_view)
            viewHolder.tvWord = mConvertView.findViewById(R.id.word_view)
            viewHolder.image = mConvertView.findViewById(R.id.image_view)

            view = mConvertView
            mConvertView.tag = viewHolder
        } else {
            viewHolder = mConvertView.tag as ViewHolder
            view = mConvertView
        }

        val animation = AnimationUtils.loadAnimation(mContext, if (position > lastPosition) R.anim.up_from_bottom else R.anim.down_from_top)
        view.startAnimation(animation)
        lastPosition = position

        viewHolder.tvPos!!.text = (position + 1).toString()
        viewHolder.tvWord!!.text = word

        return mConvertView
    }

    /*
     * ViewHolder that encapsulates the objects
     * that are used in the list view.
     */
    private class ViewHolder {
        internal var tvPos: TextView? = null
        internal var tvWord: TextView? = null
        internal var image: ImageView? = null
    }

}